File: 程序员面试金典第6版/面试题 08.06. Hanota LCCI.java
// 取巧解法无意义
// 参见下面

class Solution {
    public void hanota(List<Integer> A, List<Integer> B, List<Integer> C) {
        C.addAll(A);
    }
}

// 原地逆置栈 这道题可以让你深入了解递归

class Solution {
    public void reverseStack(Stack<Integer> stack) {
        if(stack.empty()) return;
        int top = stack.pop();
        reverseStack(stack);
        addToStackBottom(stack, top);
    }
    // 注意区分stack 和 top的区别
    public void addToStackBottom(stack, top) {
        if(stack.empty()) {
            stack.push(top);
        } else {
            int peek = stack.pop();
            addToStackBottom(stack, top);
            stack.push(peek);
        }
    }
}

File: 程序员面试金典第6版/面试题 08.06. Hanota LCCI.py
class Solution:
    def hanota(self, A: List[int], B: List[int], C: List[int]) -> None:
        """
        Do not return anything, modify C in-place instead.
        """
        n = len(A)
        self.move(n, A, B, C)
    
    def move(self, n, A, B, C):
        if n == 1:                  # 只剩最后一个盘子 A上最后一个盘子移动到C上完成
            C.append(A[-1])
            A.pop()
            return
        else:
            self.move(n-1, A, C, B) # 不是最后一个盘子 通过C将A的n-1个盘子移动到B上
            C.append(A[-1])         # 然后再将A上最后一个盘子移动到C上 同最后一步
            A.pop()
            self.move(n-1, B, A, C) # 通过C将B的n-1个盘子移动到A上 还原到操作开始状态 重复迭代